File: 0621-task-scheduler/0621-task-scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        int time = -1;
        priority_queue<int> maxHeap;
        queue<pair<int, int>> queue;    // (count_of_task, next_ready_to_schedule_time)
        vector<int> count(26, 0);

        for (auto& task : tasks) {
            count[task - 'A']++;
        }

        for (auto& c : count) {
            if (c > 0) {
                maxHeap.push(c);
            }
        }

        while (!maxHeap.empty() || !queue.empty()) {
            time++;

            if (!maxHeap.empty()) {
                int top = maxHeap.top() - 1;
                maxHeap.pop();

                if (top > 0) {
                    queue.push({top, time + n});
                }
            }

            if (!queue.empty() && queue.front().second == time) {
                maxHeap.push(queue.front().first);
                queue.pop();
            }
        }

        return time + 1;

        

        

        
        
    }
};

File: 0621-task-scheduler/0621-task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:

        count = Counter(tasks)  # creates a map with count of each variable in tasks
        maxHeap = []
        for cnt in count.values():
            maxHeap.append(-cnt)
        
        heapq.heapify(maxHeap)
        q = deque()
        time = 0
        while maxHeap or q:
            time += 1
            if maxHeap:
                cnt = 1 + heapq.heappop(maxHeap)
                if cnt:
                    q.append((cnt, time + n))
            
            if q and q[0][1] == time:
                freq = q.popleft()[0]
                heapq.heappush(maxHeap, freq)
            

        return time






        
